lib/node_modules/@stdlib/lapack/base/dgebal/lib/base.js

```diff
@@ -0,0 +1,351 @@
+*
+* @private
+* @param {string} job - indicates the operations to be performed
+* @param {NonNegativeInteger} N - number of rows/columns in matrix `A`
+* @param {Float64Array} A - input matrix to be balanced
+* @param {integer} strideA1 - stride of the first dimension of `A`
+* @param {integer} strideA2 - stride of the second dimension of `A`
+* @param {NonNegativeInteger} offsetA - starting index for `A`
+* @param {Int32Array} out - stores the first and last row/column of the balanced submatrix
+* @param {integer} strideOut - stride of `out`
+* @param {NonNegativeInteger} offsetOut - starting index for `out`
+* @param {Float64Array} scale - array containing permutation and scaling information
+* @param {integer} strideScale - stride of `scale`
+* @param {NonNegativeInteger} offsetScale  - starting index for `scale`
+* @throws {RangeError} the input matrix cannot be balanced. Encountered a NaN during computation.
+* @returns {integer} status code
+*
+* @example
+* var Float64Array = require( '@stdlib/array/float64' );
+* var Int32Array = require( '@stdlib/array/int32' );
+*
+* var A = new Float64Array( [ 1.0, 100.0, 0.0, 2.0, 200.0, 0.0, 0.0, 0.0, 3.0 ] );
+* var out = new Int32Array( 2 );
+* var scale = new Float64Array( 3 );
+*
+* dgebal( 'both', 3, A, 3, 1, 0, out, 1, 0, scale, 1, 0 );
+* // A => <Float64Array>[ 1, 12.5, 0, 16, 200, 0, 0, 0, 3 ]
+* // out => <Int32Array>[ 0, 1 ]
+* // scale => <Float64Array>[ 8, 1, 2 ]
+*/
+function dgebal( job, N, A, strideA1, strideA2, offsetA, out, strideOut, offsetOut, scale, strideScale, offsetScale ) {
+	var canSwap;
+	var noconv;
+	var ica;
+	var ira;
+	var ia1;
+	var ia2;
+	var ia3;
+	var ia4;
+	var oa1;
+	var oa2;
+	var oa3;
+	var oa4;
+	var ca;
+	var ra;
+	var is;
+	var os;
+	var c;
+	var r;
+	var k;
+	var l;
+	var i;
+	var j;
+	var g;
+	var f;
+	var s;
+
+	// Quick return if possible
+	if ( N === 0 ) {
+		out[ offsetOut ] = 0.0; // ilo
+		out[ offsetOut + strideOut ] = -1.0; // ihi (invalid)
+		return 0;
+	}
+
+	if ( job === 'none' ) {
+		dfill( N, 1.0, scale, strideScale, offsetScale );
+
+		out[ offsetOut ] = 0.0; // ilo
+		out[ offsetOut + strideOut ] = N - 1; // ihi
+		return 0;
+	}
+
+	// Permutation to isolate eigenvalues if possible
+	k = 0;
+	l = N - 1;
+
+	if ( job !== 'scale' ) {
+		// Row and column exchange
+		noconv = true;
+		while ( noconv ) {
+			// Search for rows isolating an eigenvalue and push them down
+			noconv = false;
+
+			is = offsetScale + ( l * strideScale ); // Follows scale
+			ia1 = offsetA + ( l * strideA2 ); // Follows `i`th column of A
+			ia2 = offsetA + ( l * strideA2 ); // Follows `l`th column of A
+			ia3 = offsetA + (l*strideA1); // Follows `i`th row of A
+			ia4 = offsetA + (l*strideA1); // Follows `l`th row of A
+
+			for ( i = l; i >= 0; i-- ) {
+				canSwap = true;
+				for ( j = 0; j <= l; j++ ) {
+					if ( i !== j && A[ offsetA + (i*strideA1) + (j*strideA2) ] !== 0.0 ) {
+						canSwap = false;
+						break;
+					}
+				}
```

Comment:
```suggestion
				ia = offsetA + ( i*strideA1 );
				for ( j = 0; j <= l; j++ ) {
					if ( i !== j && A[ ia ] !== 0.0 ) {
						canSwap = false;
						break;
					}
					ia += strideA2;
				}
```

Applies similarly below.